dhcp/dhcp.go

```diff
@@ -0,0 +1,461 @@
+//go:build linux
+// +build linux
+
+package dhcp
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+	"net"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/sys/unix"
+)
+
+const (
+	dhcpDiscover             = 1
+	bootRequest              = 1
+	ethPAll                  = 0x0003
+	MaxUDPReceivedPacketSize = 8192
+	dhcpServerPort           = 67
+	dhcpClientPort           = 68
+	dhcpOpCodeReply          = 2
+	bootpMinLen              = 300
+	bytesInAddress           = 4 // bytes in an ip address
+	macBytes                 = 6 // bytes in a mac address
+	udpProtocol              = 17
+
+	opRequest     = 1
+	htypeEthernet = 1
+	hlenEthernet  = 6
+	hops          = 0
+	secs          = 0
+	flags         = 0x8000 // Broadcast flag
+)
+
+// TransactionID represents a 4-byte DHCP transaction ID as defined in RFC 951,
+// Section 3.
+//
+// The TransactionID is used to match DHCP replies to their original request.
+type TransactionID [4]byte
+
+var (
+	magicCookie        = []byte{0x63, 0x82, 0x53, 0x63} // DHCP magic cookie
+	DefaultReadTimeout = 3 * time.Second
+	DefaultTimeout     = 3 * time.Second
+)
+
+type DHCP struct {
+	logger *zap.Logger
+}
+
+func New(logger *zap.Logger) *DHCP {
+	return &DHCP{
+		logger: logger,
+	}
+}
+
+type Socket struct {
```

Comment:
Looking at this Socket now, it's actually most of the way to a `net.Conn` (probably need `net.PacketConn` actually). You probably considered it, but I'd like to hear the reasoning: was there an implementation of that in the standard library that fit our needs here? If not, I'm fine with our own implementation... otherwise I'd rather use the standard library.

Reply:
no we dont have dhcp client in standard library. thats why we building our own

Reply:
I think there might be trouble with this approach because we write using one type of socket with specific settings but then read using a different type of socket with different settings-- I needed control over the socket creation itself

Reply:
Why didn't we use something like https://github.com/insomniacslk/dhcp ? It's actively maintained, used by many large projects? I get that "a little copying is better than a little dependency", but in this case where we're most of the way to implementing a full client, it feels like we should just use one. Otherwise it feels like NIH syndrome.

Reply:
I initially went with this approach, but I think we didn't want to take on the dependencies/possible license issues. Feel free to follow up with Tamilmani on this.